drivers/hm3301/hm3301.c

```diff
@@ -0,0 +1,156 @@
+#define HM3301_RESET_TIME_US       (10)
+
+int _set_i2c_mode(hm3301_t *dev)
+{
+    if (i2c_acquire(dev->params.i2c) != 0) {
+        DEBUG("[hm3301]: could not acquire I2C bus %d\n", dev->params.i2c);
+        return -1;
+    }
+    uint8_t cmd = HM3301_CMD_I2C_MODE;
+    int ret = i2c_write_bytes(dev->params.i2c, HM3301_I2C_ADDRESS, &cmd, 1, 0);
+    i2c_release(dev->params.i2c);
+
+    return ret;
+}
+
+int hm3301_init(hm3301_t *dev, const hm3301_params_t *params)
+{
+    assert(dev && params);
+    memset(dev, 0, sizeof(hm3301_t));
+    dev->params = *params;
+
+    if (gpio_is_valid(dev->params.reset_pin)) {
+        if (gpio_init(dev->params.reset_pin, GPIO_OUT)) {
+            DEBUG_PUTS("[hm3301]: failed to init reset pin");
+            return -EIO;
+        }
+    }
+
+    if (gpio_is_valid(dev->params.set_pin)) {
+        if (gpio_init(dev->params.set_pin, GPIO_OUT)) {
+            DEBUG_PUTS("[hm3301]: failed to init set pin");
+            return -EIO;
+        }
+        gpio_set(dev->params.set_pin);
+    }
+
+    DEBUG_PUTS("[hm3301]: reset device if reset_pin");
+    hm3301_reset(dev);
+
+    if (_set_i2c_mode(dev)) {
+        DEBUG_PUTS("[hm3301]: failed set i2c mode");
+        return -EPROTO;
+    }
+
+    return 0;
+}
+
+int hm3301_read(hm3301_t *dev, hm3301_data_t* data)
+{
+    if (i2c_acquire(dev->params.i2c) != 0) {
+        DEBUG("[hm3301]: could not acquire I2C bus %d\n", dev->params.i2c);
+        return -EPROTO;
+    }
+
+    uint8_t buf[HM3301_DATA_LENGTH] = {0};
+    if (i2c_read_bytes(dev->params.i2c, HM3301_I2C_ADDRESS, buf, HM3301_DATA_LENGTH, 0)) {
+        return -EPROTO;
+    }
+
+    /* calculate crc */
+    uint8_t crc = 0;
+    for (uint8_t i = 0; i < HM3301_DATA_LENGTH - 1; i++) {
+        crc += buf[i];
+    }
+
+    if (crc != buf[HM3301_DATA_LENGTH - 1]) {
+        DEBUG("crc mismatch expected %02x got %02x\n", buf[HM3301_DATA_LENGTH - 1], crc)
+    }
+
+    i2c_release(dev->params.i2c);
+
+    hm3301_data_t tmp = {
+        .mc_pm_1    = (buf[ 4] << 8) | buf[ 5],
+        .mc_pm_2p5  = (buf[ 6] << 8) | buf[ 7],
+        .mc_pm_10   = (buf[ 8] << 8) | buf[ 9],
+        .amc_pm_1   = (buf[10] << 8) | buf[11],
+        .amc_pm_2p5 = (buf[12] << 8) | buf[13],
+        .amc_pm_10  = (buf[14] << 8) | buf[15],
+        .nc_pm_0p3  = (buf[16] << 8) | buf[17],
+        .nc_pm_0p5  = (buf[18] << 8) | buf[19],
+        .nc_pm_1    = (buf[20] << 8) | buf[21],
+        .nc_pm_2p5  = (buf[22] << 8) | buf[23],
+        .nc_pm_5    = (buf[24] << 8) | buf[25],
+        .nc_pm_10   = (buf[26] << 8) | buf[27],
+    };
+    memcpy(data, &tmp, sizeof(hm3301_data_t));
+
+    return 0;
+}
+
+void hm3301_reset(hm3301_t *dev)
+{
+    if (gpio_is_valid(dev->params.reset_pin)) {
+        gpio_clear(dev->params.reset_pin);
+#if IS_USED(MODULE_ZTIMER_USEC)
+        ztimer_sleep(ZTIMER_USEC, HM3301_RESET_TIME_US);
+#else
+        for (uint32_t i = 0; i < HM3301_RESET_TIME_US * (CLOCK_CORECLOCK / US_PER_SEC) ; i++) {
```

Comment:
But still, in some (rare) cases, CLOCK_CORECLOCK might not be defined, so this won't build in the case where ztimer is not used.

Reply:
maybe also define an xtimer fallback…

Also mind you that each loop iteration is at least 3 instructions, so this will be more like `3 * HM3301_RESET_TIME_US`

Reply:
> But still, in some (rare) cases, CLOCK_CORECLOCK might not be defined, so this won't build in the case where ztimer is not used.

Currently, it's defined for all platforms, so I rather leave it like this for now. I added handling for xtimer.

Reply:
> Currently, it's defined for all platforms

Not on fe310 if HFROSC clock source is used (as explained in https://github.com/RIOT-OS/RIOT/pull/16333#discussion_r642039592). Have you read this comment ? (especially the end where I talk about a static inline function)

Reply:
Ok then I'll blacklist hifive, if you insist that much.

Reply:
Sorry that was the wrong suggestion (a petty one from my side), I see two options:

- add a dependency to `ztimer` for the driver
- add a generic `cpu_core_frequency` or `cpu_freqeuncy`, inlined to `CLOCK_CORECLOCK` unless a `HAD_CPU_FREQUENCY` flag is set. Now where could such definition be plassed?